Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/test/core/TestCertificateProcessor.kt

```diff
@@ -28,53 +29,103 @@ class TestCertificateProcessor @Inject constructor(
     private val rsaCryptography: RSACryptography,
     private val appConfigProvider: AppConfigProvider,
     private val qrCodeExtractor: DccQrCodeExtractor,
+    private val rsaKeyPairGenerator: RSAKeyPairGenerator,
 ) {
 
     /**
      * Register the public key with the server, a shortwhile later,
      * the test certificate components should be available, via [obtainCertificate].
      */
+    @Suppress("ComplexMethod")
     internal suspend fun registerPublicKey(
         data: RetrievedTestCertificate
     ): RetrievedTestCertificate {
         Timber.tag(TAG).d("registerPublicKey(cert=%s)", data)
 
         if (data is PCRCertificateData && data.labId.isNullOrBlank()) {
             Timber.tag(TAG).e("PCR certificate is missing valid labId: %s", data)
-            throw TestCertificateServerException(ErrorCode.DCC_NOT_SUPPORTED_BY_LAB)
+            throw TestCertificateException(ErrorCode.DCC_NOT_SUPPORTED_BY_LAB)
         }
 
         if (data.publicKeyRegisteredAt != null) {
             Timber.tag(TAG).d("Public key is already registered for %s", data)
             return data
         }
 
-        val publicKey = data.rsaPublicKey ?: throw IllegalArgumentException("rsaPublicKey was null")
-        requireNotNull(data.rsaPrivateKey) { "We have a public key without private key?" }
+        /*
+            https://github.com/corona-warn-app/cwa-app-android/pull/3650
+            ^ prior to the PR, a client/server error could lead to a stalemate:
+            Server says a public key was registered; App has not stored KeyPair
+            ^ the PR changes key generation to only happen once, on test registration.
+            Only users that were already affected by the bug will run into this case.
+
+            Users that upgraded right after registering for a test certificate,
+            and before retrieving a test certificate are also in this state.
+         */
+        val isMissingKeyPairEdgeCase = data.publicKeyRegisteredAt == null &&
+            data.rsaPublicKey == null &&
+            data.rsaPrivateKey == null
+
+        val (publicKey, privateKey) = when {
+            data.rsaPublicKey != null && data.rsaPrivateKey != null -> {
+                // Normal case, since ~2.5.1 (PR #3650)
+                Timber.tag(TAG).v("Using existing RSA KeyPair.")
+                data.rsaPublicKey!! to data.rsaPrivateKey!!
+            }
+            !isMissingKeyPairEdgeCase -> {
+                // Shouldn't happen
+                throw IllegalArgumentException("Missing keypair")
```

Comment:
I think the message is wrong here , right? the keyPair is not missing ,but we should be here anyway

Reply:
You are right the message doesn't fit, but unless we find something that needs to change, I'd keep it for now and not retrigger a review. It's not something that should be possible to happen anyways.

I'll fix it for 2.6.x